packages/@yoda/multimedia/mediaplayer.js

```diff
@@ -87,182 +87,124 @@ MediaPlayer.status = {
-  // TODO: nothing to to now
-}
-
-MediaPlayer.prototype.onblockpausemode = function (ext1, ext2) {
-  this.blockpausemodeEnabled = ext1 === 0
+  4: 'seekcomplete',
+  5: 'position',
+  6: 'pause',
+  7: 'stopped',
+  8: 'playing',
   /**
    * Fired when player is blocked for no cache available to play.
    * @event module:@yoda/multimedia~MediaPlayer#blockpausemode
    * @param {boolean} enabled
    */
-  this.emit('blockpausemode', this.blockpausemodeEnabled)
-}
-
-MediaPlayer.prototype.onerror = function () {
-  this._ignoreCancelEvent = true
+  9: 'blockpausemode',
+  10: 'playingstatus',
   /**
    * Fired when something went wrong.
    * @event module:@yoda/multimedia~MediaPlayer#error
    * @type {Error}
    */
-  this.emit('error', new Error('something went wrong'))
+  100: 'error',
+  200: 'info'
+}
+MediaPlayer.prototype._onevent = function (type, ext1) {
+  var eve = EventMap[type]
+  var args = [eve]
+  switch (eve) {
+    case 'prepared':
+      var vol = AudioManager.getVolume(this._stream)
+      AudioManager.setVolume(this._stream, vol)
+      this._preparing = false
+      break
+    case 'blockpausemode':
+      args.push(ext1 === 0)
+      break
+    case 'error':
+      args.push(new Error('player error'))
+      this._settled = false
+      this._preparing = false
+      break
+  }
+  this.emit.apply(this, args)
+}
+
+delegate(MediaPlayer.prototype, '_handle')
+  .method('pause')
+  .method('seekTo')
+  .method('reset')
+  .method('getAudioSessionId')
+  .method('setAudioSessionId')
+  .method('getDuration')
+  .method('getPosition')
+  .method('getIsPlaying')
+  .method('getIsLooping')
+  .method('setLooping')
+  .method('setTempoDelta')
+  .method('getVolume')
+  .method('setVolume')
+
+MediaPlayer.prototype.setDataSource = function (url) {
+  if (!this._settled) {
+    this._setup()
+  }
+  this.url = url
+  return this._handle.setDataSource(url)
 }
 
 /**
  * prepare with the given resource(URI).
  * @param {string} uri - The resource uri to play.
  * @throws {Error} uri must be a valid string.
  */
-MediaPlayer.prototype.prepare = function (uri) {
-  if (!uri) {
-    throw new Error('url must be a valid string')
+MediaPlayer.prototype.prepare = function () {
+  if (this.url && !this._settled) {
+    this.setDataSource(this.url)
   }
-  this.status = MediaPlayer.status.preparing
-  return this._handle.prepare(uri)
+  this._handle.prepare()
+  this._preparing = true
 }
 
-/**
- * start asynchronously.
- * @param {string} uri - The resource uri to play.
- * @throws {Error} uri must be a valid string.
- */
-MediaPlayer.prototype.start = function (uri) {
-  if (uri) {
-    this._handle.prepare(uri)
-    this.once('prepared', () => this._handle.start())
-    return
-  }
-  if (this.status === MediaPlayer.status.preparing) {
-    this.once('prepared', () => this._handle.start())
+MediaPlayer.prototype.start = function () {
+  if (this._preparing) {
+    this.once('prepared', () => this.start())
```

Comment:
Consider the following:

```js
var player = new MediaPlayer()
player.prepare()
for (var i = 0; i < 10; i++) {
  player.start() // this._handle.start would get called 10 times
}
```

Reply:
There are no side effects on the case since rplayer has debounced the calls.